homeassistant/components/geniushub/binary_sensor.py

```diff
@@ -17,10 +17,13 @@
     """Set up the Genius Hub sensor entities."""
     client = hass.data[DOMAIN]["client"]
 
-    devices = [d for d in client.hub.device_objs if d.type is not None]
-    switches = [
-        GeniusBinarySensor(client, d) for d in devices if d.type[:21] in GH_IS_SWITCH
-    ]
+    switches = []
+    for device in client.hub.device_objs:
+        try:
+            if device.type[:21] in GH_IS_SWITCH:
+                switches.append(GeniusBinarySensor(client, device))
+        except (AttributeError, TypeError):
```

Comment:
Why would there be these errors?

Reply:
By design, the client library returns 100% the same JSON as the vendor would via a cURL request. 

For some edge cases, `device['type'] == None`, and for others `'type' in device == False`.

The systems is based upon zwave, and one such edge case is when the devices has not been heard of for a while.  Have a look at device 4, below.

```bash
(venv) dbonnes@vm-builder:~/client_apis/geniushub-client$ python ghclient.py ${HUB_ADDRESS} -u ${USERNAME} -p ${PASSWORD} devices
WARNING:geniushubclient:Device 4, 'Room Thermostat': has no fingerprint, so has no type.
[{"id": "2", "type": "Dual Channel Receiver"}, {"id": "2-1", "type": "Dual Channel Receiver - Channel 1"}, {"id": "2-2", "type": "Dual Channel Receiver - Channel 2"}, {"id": "4"}, {"id": "6", "type": "Smart Plug"}, {"id": "8", "type": "Genius Valve"}, {"id": "9", "type": "Genius Valve"}, {"id": "10", "type": "Genius Valve"}, {"id": "11", "type": "Genius Valve"}, {"id": "13", "type": "Room Sensor"}, {"id": "14", "type": "Room Sensor"}, {"id": "15", "type": "Room Sensor"}, {"id": "16", "type": "Room Sensor"}, {"id": "17", "type": "Room Sensor"}, {"id": "19", "type": "Radiator Valve"}, {"id": "20", "type": "Radiator Valve"}, {"id": "22", "type": "Radiator Valve"}, {"id": "25", "type": "Radiator Valve"}, {"id": "27", "type": "Genius Valve"}, {"id": "29", "type": "Genius Valve"}, {"id": "30", "type": "Electric Switch"}, {"id": "34", "type": "Genius Valve"}, {"id": "36", "type": "Smart Plug"}, {"id": "38", "type": "Radiator Valve"}]

(venv) dbonnes@vm-builder:~/client_apis/geniushub-client$ curl -X GET https://my.geniushub.co.uk/v1/devices/summary -H "authorization: Bearer ${HUB_TOKEN}"
[{"id":"2","type":"Dual Channel Receiver"},{"id":"2-1","type":"Dual Channel Receiver - Channel 1"},{"id":"2-2","type":"Dual Channel Receiver - Channel 2"},{"id":"4"},{"id":"6","type":"Smart Plug"},{"id":"8","type":"Genius Valve"},{"id":"9","type":"Genius Valve"},{"id":"10","type":"Genius Valve"},{"id":"11","type":"Genius Valve"},{"id":"13","type":"Room Sensor"},{"id":"14","type":"Room Sensor"},{"id":"15","type":"Room Sensor"},{"id":"16","type":"Room Sensor"},{"id":"17","type":"Room Sensor"},{"id":"19","type":"Radiator Valve"},{"id":"20","type":"Radiator Valve"},{"id":"22","type":"Radiator Valve"},{"id":"25","type":"Radiator Valve"},{"id":"27","type":"Genius Valve"},{"id":"29","type":"Genius Valve"},{"id":"30","type":"Electric Switch"},{"id":"34","type":"Genius Valve"},{"id":"36","type":"Smart Plug"},{"id":"38","type":"Radiator Valve"}]
```

Reply:
By design, the client library returns 100% the same JSON as the vendor would via a cURL request. 

For some edge cases, `device['type'] == None`, and for others `'type' in device == False`.

The systems is based upon zwave, and one such edge case is when the devices has not been heard of for a while.  Have a look at device 4, below.

```bash
(venv) dbonnes@vm-builder:~/client_apis/geniushub-client$ python ghclient.py ${HUB_ADDRESS} -u ${USERNAME} -p ${PASSWORD} devices
WARNING:geniushubclient:Device 4, 'Room Thermostat': has no fingerprint, so has no type.
[{"id": "2", "type": "Dual Channel Receiver"}, {"id": "2-1", "type": "Dual Channel Receiver - Channel 1"}, {"id": "2-2", "type": "Dual Channel Receiver - Channel 2"}, {"id": "4"}, {"id": "6", "type": "Smart Plug"}, {"id": "8", "type": "Genius Valve"}, {"id": "9", "type": "Genius Valve"}, {"id": "10", "type": "Genius Valve"}, {"id": "11", "type": "Genius Valve"}, {"id": "13", "type": "Room Sensor"}, {"id": "14", "type": "Room Sensor"}, {"id": "15", "type": "Room Sensor"}, {"id": "16", "type": "Room Sensor"}, {"id": "17", "type": "Room Sensor"}, {"id": "19", "type": "Radiator Valve"}, {"id": "20", "type": "Radiator Valve"}, {"id": "22", "type": "Radiator Valve"}, {"id": "25", "type": "Radiator Valve"}, {"id": "27", "type": "Genius Valve"}, {"id": "29", "type": "Genius Valve"}, {"id": "30", "type": "Electric Switch"}, {"id": "34", "type": "Genius Valve"}, {"id": "36", "type": "Smart Plug"}, {"id": "38", "type": "Radiator Valve"}]

(venv) dbonnes@vm-builder:~/client_apis/geniushub-client$ curl -X GET https://my.geniushub.co.uk/v1/devices/summary -H "authorization: Bearer ${HUB_TOKEN}"
[{"id":"2","type":"Dual Channel Receiver"},{"id":"2-1","type":"Dual Channel Receiver - Channel 1"},{"id":"2-2","type":"Dual Channel Receiver - Channel 2"},{"id":"4"},{"id":"6","type":"Smart Plug"},{"id":"8","type":"Genius Valve"},{"id":"9","type":"Genius Valve"},{"id":"10","type":"Genius Valve"},{"id":"11","type":"Genius Valve"},{"id":"13","type":"Room Sensor"},{"id":"14","type":"Room Sensor"},{"id":"15","type":"Room Sensor"},{"id":"16","type":"Room Sensor"},{"id":"17","type":"Room Sensor"},{"id":"19","type":"Radiator Valve"},{"id":"20","type":"Radiator Valve"},{"id":"22","type":"Radiator Valve"},{"id":"25","type":"Radiator Valve"},{"id":"27","type":"Genius Valve"},{"id":"29","type":"Genius Valve"},{"id":"30","type":"Electric Switch"},{"id":"34","type":"Genius Valve"},{"id":"36","type":"Smart Plug"},{"id":"38","type":"Radiator Valve"}]
```

Reply:
see response in conversation.

Reply:
Hmmm...

Actually, it is a bit messy that some devices have a `self.type`, and some don't - I'm going to do something about it.... Will make `self.type = None` to stop the **AttributeError**.

Reply:
All attributes of an instance should always be initialized in init method. After init the instance should be considered complete.

TypeError is also a sign that we should do something different, either here or in the library.

Reply:
Yeah - as I was explaining it to you, I realized it was a bad idea - was just hoping I could get it past you & fix it later...

Everything is initialized in `__init__`, though.

Are you happy with `if self.type is not None`?

Reply:
> All attributes of an instance should always be initialized in init method. After init the instance should be considered complete.

How does that fit in with the `@property` decorator?

Reply:
How do you mean?

Reply:
Don't worry, I think I got it
 - `self.id` is fixed, so is set as an instance attribute in `__init__`
 - `self.name`, `self.type` are variable, so set via functions with `@property` decorator:
```python
    @property
    def type(self) -> Optional[str]:
        """Return the type of the device."""
        return self.data.get('type')
```